internal/pkg/service/buffer/worker/distribution/node.go

```diff
@@ -208,43 +221,27 @@ func (n *Node) unregister() {
 }
 
 // watch for other nodes.
-func (n *Node) watch() {
-	ctx, cancel := context.WithCancel(n.client.Ctx())
-	n.proc.OnShutdown(func() {
-		cancel()
-		n.logger.Info("cancelled watcher")
-	})
+func (n *Node) watch(ctx context.Context) {
+	pfx := n.schema.Runtime().Workers().Active().IDs()
+	ch, initDone := pfx.GetAllAndWatch(ctx, n.client, n.onWatchErr, etcd.WithPrevKV(), etcd.WithCreatedNotify())
 
+	n.wg.Add(1)
 	go func() {
+		defer n.wg.Done()
+		n.logger.Info("watching for other nodes")
 		for {
 			select {
 			case <-ctx.Done():
 				return
-			default:
-				n.logger.Infof(`watching for other nodes`)
-				pfx := n.schema.Runtime().Workers().Active().IDs()
-				ch := pfx.GetAllAndWatch(ctx, n.client, n.onWatchErr, etcd.WithPrevKV(), etcd.WithCreatedNotify())
-				n.processEvents(ctx, ch)
-
-				// Wait and try to create watcher again
-				time.Sleep(time.Second)
```

Comment:
Odtial som mohol vyhodit jedenu uroven s `for`, kedze nemusime znova vytvarat `ch` v pripade chyby, kedze to je vyriesene uz priamo vo `Watch` op, vid predchadzajuci komentar.

Teda low-level channel sa vytvori znova, a high-level channel sa nezavrie.